Trivial repl

hershen opened the pull request:
A trivial Read-Evaluate-Print Loop (REPL) which prints a prompt, reads
a line of text and prints it back out.

A CMake file is provided for building the project.

hershen committed:
Add build/ directory to .gitignore

hershen committed:
Trivial REPL

A trivial Read-Evaluate-Print Loop (REPL) which prints a prompt, reads
a line of text and prints it back out.

A CMake file is provided for building the project.

oorijsonar commented on CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+cmake_minimum_required(VERSION 3.8)
```
I think the comments in this file are not explaining anything that the code doesn't explain at a glance.

hershen replied:
I agree.

oorijsonar commented on .gitignore:
```diff
@@ -1,3 +1,6 @@
+build/
+
 # Prerequisites
 *.d
```
Not related to this pull request but no one asked me to review the initial commit :( I built the project and didn't get any `.d` files. Maybe we can clean up the `.gitignore`?

hershen replied:
That's because github automatically created it when I created the repository.

Hmm, I don't mind cleaning it up - but I'm worried that there are files that are created in Windows that I'm not aware of.
But maybe we can think of that when we start working on cross-platforming.

oorijsonar replied:
As long as I'm PM this will never be ported to windows! Unless we get paid a lot...

hershen replied:
I'll work on it on my own time, then.
Windows users deserve to be able to use good software too.

oorijsonar commented on hershensh.cpp:
```diff
@@ -0,0 +1,30 @@
+
+const std::string defaultPrompt = "hershensh$";
+
+bool isEOF() {
```
I prefer if we have the calling function above the called function. Can we do that in this project?
Also, do you have a c++ style guide that we can follow? Most projects I've seen do
```
bool isEOF()
{
<>
}
```
and not
```
bool isEOF() {
<>
}

hershen replied:
That doesn't really make sense to me - if you read the file from top to bottom, I'd expect to see the function declaration before it's usage.
Furthermore, some programming languages like python require you to define the function before using it. So I think this is a good pattern to use.

I wouldn't say there is a typical bracing pattern. See for example in the [Catch2 framework](https://github.com/catchorg/Catch2/blob/devel/src/catch2/catch_session.cpp).

I don't have a particular style guide I like, but if you find something interesting, do share.

oorijsonar replied:
No one selects a file and reads it from top to bottom. You start from the function that you're interested in and decide if you want to dive into the implementation details of the functions that it calls. If you do it the way you're doing - you will have to read going up and not going down. See the "reading code from top to bottom..." section in the chapter about functions in clean code.

oorijsonar commented on include/printer.h:
```diff
@@ -0,0 +1,11 @@
+namespace hershensh {
+void print(const std::string& line) {
+  std::cout << line << std::endl;
+}
```
This should be indented no?

hershen replied:
Debatable. I feel it won't contribute to the readability, as all code will be in the hersh namespace.
But it's not a strong opinion. What do you think?

oorijsonar replied:
[Google agrees](https://google.github.io/styleguide/cppguide.html#Namespace_Formatting).

oorijsonar commented on include/reader.h:
```diff
@@ -0,0 +1,18 @@
+
+  std::getline(std::cin, line);
+  return line;
+}
```
This should be indented no?

oorijsonar commented on CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+set(INCLUDE_DIR include)
+set(TEST_DIR tests)
+
+set(HEADER_FILES ${INCLUDE_DIR}/reader.h ${INCLUDE_DIR}/evaluate.h ${INCLUDE_DIR}/printer.h)
+
+# add the executable
+add_executable(hershensh hershensh.cpp ${HEADER_FILES})
```
This works:
```suggestion
add_executable(hershensh hershensh.cpp)
```

hershen replied:
Yes it works.
I added the headerrs so that IDEs know which files to load - see for [example](https://www.jetbrains.com/help/clion/managing-cmake-project-files.html#add-new-files).

oorijsonar replied:
I don't see them doing this in the example or explaining why this is necessary but I think the discussion is not worth the time.

oorijsonar reviewed (changes requested):
I think a standard project structure would put the source files in `src` no?

hershen committed:
Clean .gitignore

Ignoring the contents of the build directory is sufficient, as
everything is created there (at least for now).

hershen committed:
Rename hershensh to hersh

hershen committed:
Remove unnecessary comments.
